src/puppetlabs/puppetdb/http/handlers.clj

```diff
@@ -0,0 +1,385 @@
+                       (narrow-globals globals)
+                       #(s/validate catalog-query-schema
+                                    (kitchensink/mapvals sutils/parse-db-json [:edges :resources] %))
+                       (http/status-not-found-response "catalog" node)))))
+
+(defn factset-status
+  "Produces a response body for a request to retrieve the factset for the node in route-params"
+  [version]
+  (fn [{:keys [globals route-params]}]
+    (let [node (:node route-params)]
+      (status-response version
+                       ["from" "factsets" ["=" "certname" node]]
+                       (narrow-globals globals)
+                       identity
+                       (http/status-not-found-response "factset" node)))))
+
+(defn node-status
+  "Produce a response body for a single environment."
+  [version]
+  (fn [{:keys [globals route-params]}]
+    (let [node (:node route-params)]
+      (status-response version
+                       ["from" "nodes" ["=" "certname" node]]
+                       (narrow-globals globals)
+                       identity
+                       (http/status-not-found-response "node" node)))))
+
+(defn environment-status
+  "Produce a response body for a single environment."
+  [version]
+  (fn [{:keys [globals route-params]}]
+    (let [environment (:environment route-params)]
+      (status-response version
+                       ["from" "environments" ["=" "name" environment]]
+                       (narrow-globals globals)
+                       identity
+                       (http/status-not-found-response "environment" environment)))))
+
+;; Routes
+
+(pls/defn-validated events-routes :- bidi-schema/RoutePair
+  "Ring app for querying events"
+  [version :- s/Keyword]
+  (extract-query
+   {:optional (concat
+               ["query"
+                "distinct_resources"
+                "distinct_start_time"
+                "distinct_end_time"]
+               paging/query-params)}
+   (cmdi/routes
+    (cmdi/ANY "" []
+              (create-handler version "events")))))
+
+(pls/defn-validated reports-routes :- bidi-schema/RoutePair
+  [version :- s/Keyword]
+  (cmdi/routes
+   (extract-query
+    (cmdi/ANY "" []
+              (create-handler version "reports")))
+   
+   (cmdi/context ["/" :hash "/events"]
+                 (-> (events-routes version)
+                     (append-handler (comp http-q/restrict-query-to-report decode-route-params))
+                     (wrap-with-parent-check version :report :hash)))
+   
+   (cmdi/ANY ["/" :hash "/metrics"] []
+             (-> (report-data-responder version "report_metrics")
+                 (comp decode-route-params)
+                 (parent-check version :report :hash)
+                 validate-no-query-params))
+   
+   (cmdi/ANY ["/" :hash "/logs"] []
+             (-> (report-data-responder version "report_logs")
+                 (comp decode-route-params)
```

Comment:
I wasn't confident that the middleware would be applied at the right time, I can try it out though.